dapp/src/hooks/useGatingDApp.ts

```diff
@@ -0,0 +1,42 @@
+import { useEffect, useRef } from "react"
+import { MODAL_TYPES } from "#/types"
+import { featureFlags } from "#/constants"
+import { useQuery } from "@tanstack/react-query"
+import { acreApi } from "#/utils"
+import { useModal } from "./useModal"
+import useAccessCode from "./useAccessCode"
+
+export default function useGatingDApp() {
+  const { encodedCode } = useAccessCode()
+  const { openModal, closeModal, modalType } = useModal()
+  const isMounted = useRef(false)
+  const { data: isValid, isLoading } = useQuery({
+    queryKey: ["verify-access-code"],
+    enabled: !!encodedCode,
+    queryFn: async () => acreApi.verifyAccessCode(encodedCode!),
+  })
+
+  useEffect(() => {
+    if (!featureFlags.GATING_DAPP_ENABLED) return
+
+    // TODO: Add condition only for the standalone dApp
+    // if (!encodedCode && !sEmbed) return
+    if (!encodedCode) return
+
+    if (modalType === MODAL_TYPES.LOADING && !isLoading) {
+      closeModal()
+      if (!isValid) openModal(MODAL_TYPES.GATE)
+    }
+  }, [closeModal, isValid, encodedCode, isLoading, modalType, openModal])
```

Comment:
I think it should be handled inside the gating modal component. In this hook, we should only detect whether the gating modal should be displayed. Inside the gating modal, we can render the loading state, depending on the `isLoading` from the `useQuery` hook. In my opinion, the proposed solution will make the code clearer and we will get rid of the additional  `LOADING` modal type.

Reply:
I decided to split modals because `LoadingModal` and `GateModal` have different styles. Please take a look below.  Not sure if we can somehow improve it because `variant` is set in `withBaseModal`.

https://github.com/thesis/acre/blob/60d903975ff9cfa4b04023b8d2670c0ef9f135b1/dapp/src/components/LoadingModal.tsx#L22-L25

https://github.com/thesis/acre/blob/60d903975ff9cfa4b04023b8d2670c0ef9f135b1/dapp/src/components/GateModal.tsx#L48

<img width="300" alt="Screenshot 2024-10-24 at 12 52 17" src="https://github.com/user-attachments/assets/e75b9c85-6cd7-42ad-bd8c-ba43e7cd6b01">

<img width="300" alt="Screenshot 2024-10-24 at 12 52 04" src="https://github.com/user-attachments/assets/b442c2fb-1f48-47d8-a5cd-227d36065e8b">